planner/core/rule_build_key_info.go

```diff
@@ -149,6 +138,7 @@ func (p *LogicalProjection) buildSchemaByExprs(selfSchema *expression.Schema) *e
 // BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
 func (p *LogicalProjection) BuildKeyInfo(selfSchema *expression.Schema, childSchema []*expression.Schema) {
 	p.logicalSchemaProducer.BuildKeyInfo(selfSchema, childSchema)
+	selfSchema.Keys = nil
```

Comment:
Is it a bug fix?

Reply:
No.

I modify the implementation of [logicalSchemaProducer.BuildKeyInfo](https://github.com/pingcap/tidb/pull/20288/files#diff-562db6fc5a9ef67379e4fe4b795e17d2b63d27a2dcec15fd7a00153b5848e300R421), providing a default implementation for only one child logical plan, which build key info on `selfSchema` and `children[0].Schema()`.

But [LogicalProjection.BuildKeyInfo](https://github.com/pingcap/tidb/pull/20288/files#diff-b95568b5f6e1f8d798e5f14b451b7cece8e801e8a7a80567786f21ad71d46190L153) has another implementation.

```go
// A bijection exists between columns of a projection's schema and this projection's Exprs.
// Sometimes we need a schema made by expr of Exprs to convert a column in child's schema to a column in this projection's Schema.
func (p *LogicalProjection) buildSchemaByExprs(selfSchema *expression.Schema) *expression.Schema {
   ...
}

// BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
func (p *LogicalProjection) BuildKeyInfo(selfSchema *expression.Schema, childSchema []*expression.Schema) {
    ...
    schema := p.buildSchemaByExprs(selfSchema)
    ...
```

So I `selfSchema.Keys = nil` to get `selfSchema.Keys` ready for following [codes](https://github.com/pingcap/tidb/blob/38ecab6a7c1dde3c9a2ed2713b3a69d63f662b77/planner/core/rule_build_key_info.go#L163).

Reply:
How about just calculating `Keys` once with these code?
```go
func (p *LogicalProjection) BuildKeyInfo(selfSchema *expression.Schema, childSchema []*expression.Schema) {
    schema := p.buildSchemaByExprs(selfSchema)
    p.logicalSchemaProducer.BuildKeyInfo(schema, childSchema)
    copy(selfSchema.Keys, schema.Keys)
}
```

Reply:
It's not working, because the build key info implementation is special.
It gets indices from schema from `buildSchemaByExprs`, but gets key columns from `selfSchema`.

```go
	for _, key := range childSchema[0].Keys {
		indices := schema.ColumnsIndices(key) //  <-- schema
		if indices == nil {
			continue
		}
		newKey := make([]*expression.Column, 0, len(key))
		for _, i := range indices {
			newKey = append(newKey, selfSchema.Columns[i])  // <-- selfSchema
		}
		selfSchema.Keys = append(selfSchema.Keys, newKey)
	}
```

Reply:
How about [this](https://github.com/pingcap/tidb/pull/20288/commits/0abb2a4511a89951b099a3307ed776cebc31db4a#diff-b95568b5f6e1f8d798e5f14b451b7cece8e801e8a7a80567786f21ad71d46190R141) ?

```go
// BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
func (p *LogicalProjection) BuildKeyInfo(selfSchema *expression.Schema, childSchema []*expression.Schema) {
	// `LogicalProjection` use schema from `Exprs` to build key info. See `buildSchemaByExprs`.
	// So call `baseLogicalPlan.BuildKeyInfo` here to avoid duplicated building key info.
	p.baseLogicalPlan.BuildKeyInfo(selfSchema, childSchema)
	selfSchema.Keys = nil
	schema := p.buildSchemaByExprs(selfSchema)
	for _, key := range childSchema[0].Keys {
		indices := schema.ColumnsIndices(key)
		if indices == nil {
			continue
		}
		newKey := make([]*expression.Column, 0, len(key))
		for _, i := range indices {
			newKey = append(newKey, selfSchema.Columns[i])
		}
		selfSchema.Keys = append(selfSchema.Keys, newKey)
	}
}
```